src/services/anchor-service.ts

```diff
@@ -78,74 +76,90 @@ class IpfsLeafCompare implements CompareFunction<Candidate> {
+    @inject('ceramicService') private ceramicService?: CeramicService,
+    @inject('anchorRepository') private anchorRepository?: AnchorRepository) {
+
+    this.ipfsMerge = new IpfsMerge(this.ipfsService);
     this.ipfsCompare = new IpfsLeafCompare();
   }
 
-  /**
-   * Sets dependencies
-   * @param context - Application context
-   */
-  setContext(context: Context): void {
-    this.blockchainService = context.getSelectedBlockchainService();
-
-    this.ceramicService = context.lookup('CeramicService');
-    this.requestService = context.lookup('RequestService');
-    this.anchorRepository = context.lookup('AnchorRepository')
-  }
-
   /**
    * Finds anchor by request
    * @param request - Request instance
    */
   public async findByRequest(request: Request): Promise<Anchor> {
-    return this.anchorRepository.findByRequest(request)
+    return this.anchorRepository.findByRequest(request);
   }
 
   /**
    * Creates anchors for client requests
    */
   public async anchorRequests(): Promise<void> {
-    const reqs: Request[] = await this.requestService.findNextToProcess();
-    if (reqs.length === 0) {
-      logger.Info('No pending CID requests found. Skipping anchor.');
+    let requests: Request[] = await this.requestService.findNextToProcess();
+    if (requests.length === 0) {
+      logger.Info("No pending CID requests found. Skipping anchor.");
       return;
     }
 
-    // filter old updates for same docIds
-    let candidates: Candidate[] = await this._findCandidates(reqs);
-    const validReqIds = candidates.map(p => p.reqId);
-    const discardedReqs = reqs.filter(r => !validReqIds.includes(r.id));
+    const nonReachableRequestIds = await this.ipfsService.findUnreachableCids(requests);
+    if (nonReachableRequestIds.length !== 0) {
+      logger.Err("Some of the records will be discarded since they cannot be retrieved.");
+      // discard non reachable ones
+      await this.requestService.updateRequests({
+        status: RS.FAILED,
+        message: "Request has failed. Record is not reachable by CAS IPFS service."
+      }, nonReachableRequestIds);
+    }
 
-    if (discardedReqs.length > 0) {
-      // update discarded requests
-      await this.requestService.updateReqs({ status: RS.FAILED, message: 'Request has failed. There are conflicts with other requests for the same document and DID.', }, discardedReqs.map(r => r.id));
+    // filter valid requests
+    requests = requests.filter(r => !nonReachableRequestIds.includes(r.id));
+    if (requests.length === 0) {
+      logger.Info("No CID to request. Skipping anchor.");
+      return;
     }
 
-    // create merkle tree
-    const merkleTree: MerkleTree<Candidate> = await this._createMerkleTree(candidates);
-    candidates = merkleTree.getLeaves();
+    let candidates: Candidate[] = await this._findCandidates(requests);
+    const clashingRequestIds = requests.filter(r => !candidates.map(p => p.reqId).includes(r.id)).map(r => r.id);
```

Comment:
nit: use 'c' instead of 'p' for the variable in the candidates.map() lambda?

Reply:
yep, tnx

Reply:
I guess I have to rename `pair` with `candidate` in other places too 👍